### PR TITLE
FSF-2669, FSF-2667, FSF-2188 Master state logic

### PR DESCRIPTION
**Pull request by yvesb04**
various issues fixed, completed state logic actuators

### CODE REVIEW

File: master/include/logic/stateLogic.hpp [hunk 1]
```diff
@@ -1,104 +1,109 @@
...
+
+            DigitalSender::enterOffState();
             state = AS_OFF;
             break;
         default:
             break;
     }
 }
+
+inline void ASState::performEmergencyOperations() {
```
Comment (marhcouto) [inline master/include/logic/stateLogic.hpp:105]: What is this function for?

**Review by marhcouto — CHANGES_REQUESTED** [review]
Overall good job! Just missing doccumentation, which is important

File: master/include/logic/stateLogic.hpp (hunk 1 above)
Comment (yvesb04) [reply master/include/logic/stateLogic.hpp:105]: avoiding repeating the emergency operations 3 times, just call the function

**Comment by yvesb04** [comment]
> Overall good job! Just missing doccumentation, which is important

will add

File: master/platformio.ini [hunk 2]
```diff
@@ -4,15 +4,17 @@ default_envs = teensy41, teensy41-debug
 [env:teensy41]
 platform = teensy
 build_flags = -D EMBEDDED
-board = teensy41
+;board = teensy41
+board = teensy40
```
Comment (andrelds11) [inline master/platformio.ini:8]: why the change here, shouldn't it be another environment?

File: master/include/embedded/digitalReceiver.hpp [hunk 3]
```diff
@@ -4,126 +4,133 @@
...
-  // Enum value attributed considering the True Boolean Value
-  *mission = static_cast<Mission>(
-      digitalRead(MISSION_MANUAL_PIN) * MANUAL |
-      digitalRead(MISSION_ACCELERATION_PIN) * ACCELERATION |
-      digitalRead(MISSION_SKIDPAD_PIN) * SKIDPAD |
-      digitalRead(MISSION_AUTOCROSS_PIN) * AUTOCROSS |
-      digitalRead(MISSION_TRACKDRIVE_PIN) * TRACKDRIVE |
-      digitalRead(MISSION_EBSTEST_PIN) * EBS_TEST |
-      digitalRead(MISSION_INSPECTION_PIN) * INSPECTION);
+inline void DigitalReceiver::readMission() const {
```
Comment (andrelds11) [inline master/include/embedded/digitalReceiver.hpp:92]: why the consts in these functions? Using it in a read/set function may give the false idea that this is not changing any variable

File: master/include/logic/checkupManager.hpp [hunk 4]
```diff
@@ -112,10 +201,20 @@ inline bool CheckupManager::r2dCheckup() const {
 }
 
 inline bool CheckupManager::emergencyCheckup() const {
-    //TODO Continuous monitoring sequence
+    if (_systemData->failureDetection.hasAnyComponentTimedOut()
```
Comment (marhcouto) [inline master/include/logic/checkupManager.hpp:204]: Isn't the EBS check missing? Like the pneumatic sensors?

File: master/include/logic/checkupManager.hpp [hunk 5]
```diff
@@ -112,10 +201,20 @@ inline bool CheckupManager::r2dCheckup() const {
...
+        || _systemData->failureDetection.emergencySignal || _systemData->sdcState_OPEN) {
+        return EXIT_FAILURE;
+    }
     return EXIT_SUCCESS;
 }
 
+inline bool CheckupManager::drivingCheckup() const {
```
Comment (marhcouto) [inline master/include/logic/checkupManager.hpp:211]: I am confused by this

File: master/include/logic/stateLogic.hpp [hunk 6]
```diff
@@ -1,104 +1,108 @@
...
                 state = AS_EMERGENCY;
                 break;
             }
-            if (_checkupManager.missionFinishedCheckup()) {
+            if (_checkupManager.missionFinishedCheckup())
                 break;
-            }
-        // TODO: perform necessary actions to enter AS_OFF
+
+            DigitalSender::enterOffState();
```
Comment (marhcouto) [inline master/include/logic/stateLogic.hpp:88]: The way these if statements with breaks are made are quite confusing

File: master/include/logic/stateLogic.hpp [hunk 7]
```diff
@@ -1,104 +1,108 @@
...
             if (_checkupManager.emergencyCheckup()) {
-                // Emergency, go to emergency state
+                performEmergencyOperations();
                 state = AS_EMERGENCY;
                 break;
             }
-            if (_checkupManager.missionFinishedCheckup()) {
-                break;
-            }
+            if (_checkupManager.drivingCheckup()) break;
```
Comment (marhcouto) [inline master/include/logic/stateLogic.hpp:74]: I think the combination of having a successfull exit as a 0 (which is typical in some contexts but may be unexpected in this one, as usually it is coupled with an error check) and the fact that you use the break instead of reversing the logic and puting the rest inside of the if statement leaves this code more confusing than it should be.

File: master/include/logic/stateLogic.hpp (hunk 1 above)
Comment (marhcouto) [reply master/include/logic/stateLogic.hpp:104]: but there is only one statement inside of it

File: master/include/embedded/digitalSender.hpp [hunk 8]
```diff
@@ -3,6 +3,132 @@
...
+    static void enterReadyState();
+
+    void enterDrivingState();
+
+    static void enterFinishState();
+
+    void blinkLED();
 };
+
+inline void DigitalSender::sendDigitalSignal(const int pin, const int signal) {
```
Comment (marhcouto) [inline master/include/embedded/digitalSender.hpp:56]: What is this for?

**Review by marhcouto — COMMENTED** [review]
OVerall great job at making organized code, just a last suggestion: the state machine could be further simplified if you just acted the states functions everytime in their case, instead of doing it once before you enter them

**Review by marhcouto — CHANGES_REQUESTED** [review]
Well those are the changes requested

**Review by marhcouto — COMMENTED** [review]
Sorry for taking so long

File: master/platformio.ini (hunk 2 above)
Comment (yvesb04) [reply master/platformio.ini:8]: yes i pushed the wrong thing, for testing we need to use teensy40 as that's the breadboard testing board

File: master/include/embedded/digitalReceiver.hpp (hunk 3 above)
Comment (yvesb04) [reply master/include/embedded/digitalReceiver.hpp:92]: clion suggestions, they can be const as they are not editing any class attributes but rather (the pointer adress i.e), but we can remove this

File: master/include/embedded/digitalSender.hpp (hunk 8 above)
Comment (yvesb04) [reply master/include/embedded/digitalSender.hpp:56]: buzzer signal

**Comment by yvesb04** [comment]
Everything fixed I believe, please re-review. Missing correct brake pressure checks somewhere (probably), please follow the state logic sequence and tell me where so I can fix this. Implemented watchdog correctly too. removed consts, renamed everything for easy to follow logic

**Comment by andrelds11** [comment]
> Everything fixed I believe, please re-review. Missing correct brake pressure checks somewhere (probably), please follow the state logic sequence and tell me where so I can fix this. Implemented watchdog correctly too. removed consts, renamed everything for easy to follow logic

I think the brake pressure is right, your checking it in both manual and as off states

File: master/include/logic/systemDiagnostics.hpp [hunk 9]
```diff
@@ -35,19 +37,19 @@ struct InternalLogics {
 };
 
 struct FailureDetection {
-    Timestamp pcAliveTimestamp, steerAliveTimestamp, inversorAliveTimestamp,
-            bmsAliveTimestamp;
+    Metro pcAliveTimestamp{COMPONENT_TIMESTAMP_TIMEOUT};
+    Metro steerAliveTimestamp{COMPONENT_TIMESTAMP_TIMEOUT};
+    Metro inversorAliveTimestamp{COMPONENT_TIMESTAMP_TIMEOUT};
+    Metro bmsAliveTimestamp{COMPONENT_TIMESTAMP_TIMEOUT};
```
Comment (marhcouto) [inline master/include/logic/systemDiagnostics.hpp:43]: To remove

File: master/include/logic/checkupManager.hpp [hunk 10]
```diff
@@ -68,75 +94,179 @@ class CheckupManager {
...
+    _systemData->digitalData.watchdogTimestamp.reset();
+    return false;
 }
 
-inline bool CheckupManager::emergencyCheckup() const {
-    //TODO Continuous monitoring sequence
-    return EXIT_SUCCESS;
+inline bool CheckupManager::shouldEnterEmergency() const {
+    if (_systemData->failureDetection.hasAnyComponentTimedOut() ||
+        _systemData->failureDetection.emergencySignal ||
```
Comment (marhcouto) [inline master/include/logic/checkupManager.hpp:236]: Missing check to pneumatic sensors
```suggestion
        _systemData->digitalData.pneumatic_line_pressure || 
```

File: master/include/logic/checkupManager.hpp [hunk 11]
```diff
@@ -68,75 +94,179 @@ class CheckupManager {
...
+            initialCheckupTimestamp.reset();
+            state = CheckupState::CHECK_WATCHDOG;
+            break;
+        case CheckupState::CHECK_WATCHDOG:
+            // Watchdog_is_ready == 0
+            if (initialCheckupTimestamp.check() && !_systemData->digitalData.watchdog_state) {
+                state = CheckupState::CLOSE_SDC;
+                //Start toggling watchdog again (will only actually happen when the car is in AS_READY)
+            }
+            break;
```
Comment (marhcouto) [inline master/include/logic/checkupManager.hpp:171]: Needs to start toggling watchdog again for the next states

File: master/include/logic/checkupManager.hpp [hunk 12]
```diff
@@ -68,75 +94,179 @@ class CheckupManager {
...
+    return false;
 }
 
-inline bool CheckupManager::emergencyCheckup() const {
-    //TODO Continuous monitoring sequence
-    return EXIT_SUCCESS;
+inline bool CheckupManager::shouldEnterEmergency() const {
+    if (_systemData->failureDetection.hasAnyComponentTimedOut() ||
+        _systemData->failureDetection.emergencySignal ||
+        _systemData->sdcState_OPEN ||
```
Comment (marhcouto) [inline master/include/logic/checkupManager.hpp:237]: Needs to check if ASMS is off or if hydraulic brakes are active for AS Driving as well

File: master/include/logic/systemDiagnostics.hpp [hunk 13]
```diff
@@ -35,19 +37,19 @@ struct InternalLogics {
 };
 
 struct FailureDetection {
-    Timestamp pcAliveTimestamp, steerAliveTimestamp, inversorAliveTimestamp,
-            bmsAliveTimestamp;
+    Metro pcAliveTimestamp{COMPONENT_TIMESTAMP_TIMEOUT};
+    Metro steerAliveTimestamp{COMPONENT_TIMESTAMP_TIMEOUT};
+    Metro inversorAliveTimestamp{COMPONENT_TIMESTAMP_TIMEOUT};
```
Comment (marhcouto) [inline master/include/logic/systemDiagnostics.hpp:42]: Missing updates to this timestamp

File: master/include/logic/checkupManager.hpp [hunk 14]
```diff
@@ -68,75 +94,179 @@ class CheckupManager {
...
+
+    _systemData->digitalData.watchdogTimestamp.reset();
+    return false;
 }
 
-inline bool CheckupManager::emergencyCheckup() const {
-    //TODO Continuous monitoring sequence
-    return EXIT_SUCCESS;
+inline bool CheckupManager::shouldEnterEmergency() const {
+    if (_systemData->failureDetection.hasAnyComponentTimedOut() ||
```
Comment (marhcouto) [inline master/include/logic/checkupManager.hpp:235]: Also needs to check if TS is on (message from bamocar telling the tension)

**Review by marhcouto — CHANGES_REQUESTED** [review]
Make changes and then merge right away

File: master/include/logic/systemDiagnostics.hpp (hunk 9 above)
Comment (marhcouto) [reply master/include/logic/systemDiagnostics.hpp]: Just the bms one
